src/java/org/apache/cassandra/security/PEMBasedSslContextFactory.java

```diff
@@ -90,47 +90,55 @@ public final class PEMBasedSslContextFactory extends FileBasedSslContextFactory
 {
     public static final String DEFAULT_TARGET_STORETYPE = "PKCS12";
     private static final Logger logger = LoggerFactory.getLogger(PEMBasedSslContextFactory.class);
-    private String pemEncodedKey;
-    private String keyPassword;
-    private String pemEncodedCertificates;
-    private boolean maybeFileBasedPrivateKey;
-    private boolean maybeFileBasedTrustedCertificates;
+    private PEMBasedKeyStoreContext pemEncodedTrustCertificates;
+    private PEMBasedKeyStoreContext pemEncodedKeyContext;
+    private PEMBasedKeyStoreContext pemEncodedOutboundKeyContext;
 
     public PEMBasedSslContextFactory()
     {
     }
 
-    public PEMBasedSslContextFactory(Map<String, Object> parameters)
+    private void validatePasswords()
     {
-        super(parameters);
-        pemEncodedKey = getString(ConfigKey.ENCODED_KEY.getKeyName());
-        keyPassword = getString(ConfigKey.KEY_PASSWORD.getKeyName());
-        if (StringUtils.isEmpty(keyPassword))
-        {
-            keyPassword = keystore_password;
-        }
-        else if (!StringUtils.isEmpty(keystore_password) && !keyPassword.equals(keystore_password))
+        boolean shouldThrow = !keystoreContext.passwordMatchesIfPresent(pemEncodedKeyContext.password)
+                              || !outboundKeystoreContext.passwordMatchesIfPresent(pemEncodedOutboundKeyContext.password);
+        boolean isOutbound = !outboundKeystoreContext.passwordMatchesIfPresent(pemEncodedOutboundKeyContext.password);
```

Comment:
It took me a moment to understand why this was the check for outbound. You could DRY this up by renaming `isOutbound` to something more accurate and reusing it.
```suggestion
        boolean outboundPasswordMismatch = !outboundKeystoreContext.passwordMatchesIfPresent(pemEncodedOutboundKeyContext.password);
        boolean shouldThrow = !keystoreContext.passwordMatchesIfPresent(pemEncodedKeyContext.password)
                              || outboundPasswordMismatch;
        String keyPrefix = outboundPasswordMismatch ? "outbound_" : "";
```